beluga/include/beluga/algorithm/spatial_hash.hpp

```diff
@@ -196,6 +196,61 @@ class spatial_hash<Sophus::SE2d, void> {
   spatial_hash<std::tuple<double, double, double>> underlying_hasher_{{1., 1., 1.}};
 };
 
+/**
+ * Specialization for Sophus::SE3d. Will calculate the spatial hash based on the translation and rotation (expressed in
+ * Euler Angles).
+ */
+template <>
+class spatial_hash<Sophus::SE3d, void> {
+ public:
+  /// Constructs a spatial hasher given per-coordinate resolutions.
+  /**
+   * \param x_clustering_resolution Clustering resolution for the X axis, in meters.
+   * \param y_clustering_resolution Clustering resolution for the Y axis, in meters.
+   * \param z_clustering_resolution Clustering resolution for the Z axis, in meters.
+   * \param angular_clustering_resolution Clustering resolution for roll pitch and yaw, in radians.
+   */
+  explicit spatial_hash(
+      double x_clustering_resolution,
+      double y_clustering_resolution,
+      double z_clustering_resolution,
+      double angular_clustering_resolution)
+      : underlying_hasher_{
+            {x_clustering_resolution, y_clustering_resolution, z_clustering_resolution, angular_clustering_resolution,
+             angular_clustering_resolution, angular_clustering_resolution}} {};
+
+  /// Constructs a spatial hasher given per-group resolutions.
+  /**
+   * \param linear_clustering_resolution Clustering resolution for translational coordinates, in meters.
+   * \param angular_clustering_resolution Clustering resolution for rotational components, in radians.
+   */
+  explicit spatial_hash(double linear_clustering_resolution, double angular_clustering_resolution)
+      : spatial_hash(
+            linear_clustering_resolution,
+            linear_clustering_resolution,
+            linear_clustering_resolution,
+            angular_clustering_resolution){};
+
+  /// Default constructor
+  spatial_hash() = default;
+
+  /// Calculates the tuple hash, using the given resolution for x, y, z  and rotation given at construction time.
+  /**
+   * \param state The state to be hashed.
+   * \return The calculated hash.
+   */
+  std::size_t operator()(const Sophus::SE3d& state) const {
+    const auto& position = state.translation();
+    // This gives yaw, pitch, roll in that order.
+    const Eigen::Vector3d euler_angles = state.so3().matrix().eulerAngles(2, 1, 0);
+    return underlying_hasher_(std::make_tuple(
+        position.x(), position.y(), position.z(), euler_angles.z(), euler_angles.y(), euler_angles.x()));
```

Comment:
nit: Isn't this the same?

```suggestion
    const Eigen::Vector3d euler_angles = state.so3().matrix().eulerAngles(0, 1, 2);
    return underlying_hasher_(std::make_tuple(
        position.x(), position.y(), position.z(), euler_angles.x(), euler_angles.y(), euler_angles.z()));
```

Reply:
Added a bit of context on why this is not the same I'm going for. LMK if that's clear enough